datalad/support/network.py

```diff
@@ -235,3 +249,503 @@ def __call__(self, url):
             return dict(mtime=r_stamp['mtime'], size=r_stamp['size'], url=url)
         finally:
             r.close()
+
+
+# TODO:  make it consistent/clear at what stage % encoding/decoding happens!
+# now it is a mix!
+
+#
+# Useful functionality in requests.models
+#  utils.requote_uri -- quote/unquote cycle to guarantee consistent appearance
+#  RequestEncodingMixin._encode_params -- Will successfully encode parameters when passed as a dict or a list of ...
+#  PreparedRequest().prepare_url(url, params) -- nicely cares about url encodings etc
+#
+
+def _guess_ri_cls(ri):
+    """Factory function which would determine which type of a ri a provided string is"""
+    # We assume that it is a URL and parse it. Depending on the result
+    # we might decide that it was something else ;)
```

Comment:
it might indeed not scale and indeed I think we might want `is_valid`.  But unless you have specific examples to address right away, I would prefer to work on that later and get this one merged, although I am still on the fence either to use '#' (implemented in this PR) or '?' for dl+archive urls as discussed in https://github.com/datalad/datalad/issues/441#issuecomment-223376906